server/src/main/java/org/elasticsearch/action/bulk/TransportAbstractBulkAction.java

```diff
@@ -180,12 +181,34 @@ protected void doRun() throws IOException {
     private boolean applyPipelines(Task task, BulkRequest bulkRequest, Executor executor, ActionListener<BulkResponse> listener)
         throws IOException {
         boolean hasIndexRequestsWithPipelines = false;
-        final Metadata metadata = clusterService.state().getMetadata();
-        Map<String, ComponentTemplate> templateSubstitutions = bulkRequest.getComponentTemplateSubstitutions();
+        final Metadata metadata;
+        Map<String, ComponentTemplate> componentTemplateSubstitutions = bulkRequest.getComponentTemplateSubstitutions();
+        if (bulkRequest.isSimulated() && componentTemplateSubstitutions.isEmpty() == false) {
+            /*
+             * If this is a simulated request, and there are template substitutions, then we want to create and use a new metadata that has
+             * those templates.
+             */
```

Comment:
I think it will be really helpful to refer here to the methods that require this to work properly.

Reply:
I added a bit here. Let me know if there's something else that would be helpful.